File: src/operator/contrib/tvmop/ufunc.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * Copyright (c) 2019 by Contributors
 * \file ufunc.cc
 * \brief
 * \author Yizhi Liu
 */
#ifdef MXNET_USE_TVM_OP
#include <tvm/runtime/packed_func.h>
#include <tvm/runtime/registry.h>
#include <tvm/runtime/c_runtime_api.h>
#include <mxnet/base.h>
#include "../../tensor/elemwise_binary_broadcast_op.h"
#include "../../tvmop/op_module.h"
#include "../../tensor/elemwise_binary_op.h"

namespace mxnet {
namespace op {

static constexpr char func_vadd_cpu[] = "vadd";
static constexpr char func_vadd_gpu[] = "cuda_vadd";

template<const char* func>
void TVMBroadcastCompute(const nnvm::NodeAttrs& attrs,
                         const mxnet::OpContext& ctx,
                         const std::vector<TBlob>& inputs,
                         const std::vector<OpReqType>& req,
                         const std::vector<TBlob>& outputs) {
  CHECK_EQ(inputs.size(), 2U);
  CHECK_EQ(outputs.size(), 1U);
  tvm::runtime::TVMOpModule::Get()->Call(func, ctx, {inputs[0], inputs[1], outputs[0]});
}

NNVM_REGISTER_OP(_contrib_tvm_vadd)
    .set_num_inputs(2)
    .set_num_outputs(1)
    .add_argument("a", "NDArray-or-Symbol", "first input")
    .add_argument("b", "NDArray-or-Symbol", "second input")
    .set_attr<mxnet::FInferShape>("FInferShape", BinaryBroadcastShape)
    .set_attr<nnvm::FInferType>("FInferType", mxnet::op::ElemwiseType<2, 1>)
    .set_attr<mxnet::FCompute>("FCompute<cpu>", mxnet::op::TVMBroadcastCompute<func_vadd_cpu>)
#if MXNET_USE_CUDA
    .set_attr<mxnet::FCompute>("FCompute<gpu>", mxnet::op::TVMBroadcastCompute<func_vadd_gpu>);
#endif  // MXNET_USE_CUDA

static constexpr char func_exp2_cpu[] = "exp2";
static constexpr char func_epx2_gpu[] = "cuda_exp2";
static constexpr char func_backward_epx2_cpu[] = "backward_exp2";
static constexpr char func_backward_epx2_gpu[] = "cuda_backward_exp2";

template<const char* func>
void TVMOpExp2Compute(const nnvm::NodeAttrs& attrs,
                      const mxnet::OpContext& ctx,
                      const std::vector<TBlob>& inputs,
                      const std::vector<OpReqType>& req,
                      const std::vector<TBlob>& outputs) {
  CHECK_EQ(inputs.size(), 1U);
  CHECK_EQ(outputs.size(), 1U);
  tvm::runtime::TVMOpModule::Get()->Call(func, ctx, {inputs[0], outputs[0]});
}

//template<int req>
//struct exp2_backward {
//  template<typename DType>
//  MSHADOW_XINLINE static void Map(int i, DType* in_grad, const DType* in_data1,
//                                  const DType* in_data2, const DType* out_grad) {
//    DType grad = in_data2[i] * in_data1[i] / DType(2);
//    KERNEL_ASSIGN(in_grad[i], req, grad);
//  }
//};

template<const char* func>
void TVMExp2Backward(const nnvm::NodeAttrs& attrs,
                     const OpContext& ctx,
                     const std::vector<TBlob>& inputs,
                     const std::vector<OpReqType>& req,
                     const std::vector<TBlob>& outputs) {
  CHECK_EQ(inputs.size(), 3U);
  CHECK_EQ(outputs.size(), 1U);
  CHECK_EQ(req.size(), 1U);
  using namespace mshadow;
  const TBlob& out_grad = inputs[0];
  const TBlob& in_data1 = inputs[1];
  const TBlob& in_data2 = inputs[2];
  const TBlob& in_grad = outputs[0];
  tvm::runtime::TVMOpModule::Get()->Call(func, ctx, {out_grad, in_data1, in_data2, in_grad});
}

NNVM_REGISTER_OP(_contrib_tvm_exp2)
    .set_num_inputs(1)
    .set_num_outputs(1)
    .add_argument("data", "NDArray-or-Symbol", "resources")
    .set_attr<mxnet::FInferShape>("FInferShape", mxnet::op::ElemwiseShape<1, 1>)
    .set_attr<nnvm::FInferType>("FInferType", mxnet::op::ElemwiseType<1, 1>)
#if MXNET_USE_CUDA
    .set_attr<mxnet::FCompute>("FCompute<gpu>", mxnet::op::TVMOpExp2Compute<func_epx2_gpu>)
#endif  // MXNET_USE_CUDA
    .set_attr<mxnet::FCompute>("FCompute<cpu>", mxnet::op::TVMOpExp2Compute<func_exp2_cpu>)
    .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseInOut{ "_tvm_backward_exp2" });



NNVM_REGISTER_OP(_tvm_backward_exp2)
.set_num_inputs(3)
.set_num_outputs(1)
.set_attr<nnvm::TIsBackward>("TIsBackward", true)
#if MXNET_USE_CUDA
.set_attr<FCompute>("FCompute<gpu>", mxnet::op::TVMExp2Backward<func_backward_epx2_gpu>)
#endif  // MXNET_USE_CUDA
.set_attr<FCompute>("FCompute<cpu>", mxnet::op::TVMExp2Backward<func_backward_epx2_cpu>);

}  // namespace op
}  // namespace mxnet
#endif  // MXNET_USE_TVM_OP

File: contrib/tvmop/basic/ufunc.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

# coding: utf-8
import tvm
import topi
from .. import defop, AllTypes

def compute_add(dtype, ndim):
    A = tvm.placeholder([tvm.var() for _ in range(ndim)], name='A', dtype=dtype)
    B = tvm.placeholder([tvm.var() for _ in range(ndim)], name='B', dtype=dtype)
    C = tvm.compute([tvm.var() for _ in range(ndim)],
                    lambda *index: A[index] + B[index], name='C')
    s = tvm.create_schedule(C.op)
    return s, A, B, C

@defop(name="vadd", target="cpu", auto_broadcast=True,
       dtype=AllTypes, ndim=list(range(1, 6)))
def vadd(dtype, ndim):
    s, A, B, C = compute_add(dtype, ndim)
    axes = [axis for axis in C.op.axis]
    fused = s[C].fuse(*axes)
    s[C].parallel(fused)

    return s, [A, B, C]

@defop(name="cuda_vadd", target="cuda", auto_broadcast=True,
       dtype=["float32", "float64"], ndim=list(range(1, 6)))
def vadd_gpu(dtype, ndim):
    s, A, B, C = compute_add(dtype, ndim)
    s = tvm.create_schedule(C.op)
    axes = [axis for axis in C.op.axis]
    fused = s[C].fuse(*axes)
    bx, tx = s[C].split(fused, factor=64)
    s[C].bind(bx, tvm.thread_axis("blockIdx.x"))
    s[C].bind(tx, tvm.thread_axis("threadIdx.x"))
    return s, [A, B, C]

def compute_exp2(dtype, ndim):
    A = tvm.placeholder([tvm.var() for _ in range(ndim)], name='A', dtype=dtype)
    B = tvm.compute([tvm.var() for _ in range(ndim)],
                    lambda *index: topi.power(2, A[index]), name='B')
    s = tvm.create_schedule(B.op)
    return s, A, B

@defop(name="exp2", target="cpu", auto_broadcast=True,
       dtype=["float32", "float64"], ndim=list(range(0, 6)))
def exp2(dtype, ndim):
    s, A, B = compute_exp2(dtype, ndim)
    # axes = [axis for axis in B.op.axis]
    # fused = s[B].fuse(*axes)
    # opt0
    # s[B].parallel(fused)
    # opt 1
    # fused = s[B].fuse(*axes)
    #
    #
    # s[B].reorder(fused)
    # s[B].parallel(fused)
    # opt 2
    # fused = s[B].fuse(*axes)
    # xo, xi = s[B].split(fused, factor=32)
    # s[B].reorder(xo, xi)
    #
    # opt 3
    # fused = s[B].fuse(*axes)
    # xo, xi = s[B].split(fused, factor=32)
    # s[B].reorder(xo, xi)
    # s[B].vectorize(xi)
    #
    # opt 4
    # fused = s[B].fuse(*axes)
    # xo, xi = s[B].split(fused, factor=32)
    # s[B].reorder(xo, xi)
    # s[B].vectorize(xi)
    # s[B].parallel(xo)

    return s, [A, B]

@defop(name="cuda_exp2", target="cuda", auto_broadcast=True,
       dtype=["float32", "float64"], ndim=list(range(0, 6)))
def exp2_gpu(dtype, ndim):
    s, A, B= compute_exp2(dtype, ndim)
    s = tvm.create_schedule(B.op)
    #opt0
    # axes = [axis for axis in B.op.axis]
    # fused = s[B].fuse(*axes)
    # bx, tx = s[B].split(fused, factor=64)
    # s[B].bind(bx, tvm.thread_axis("blockIdx.x"))
    # s[B].bind(tx, tvm.thread_axis("threadIdx.x"))
    #
    #opt1
    axes = [axis for axis in B.op.axis]
    fused = s[B].fuse(*axes)
    bx, tx = s[B].split(fused, factor=64)
    s[B].reorder(bx, tx)

    s[B].bind(bx, tvm.thread_axis("blockIdx.x"))
    s[B].bind(tx, tvm.thread_axis("threadIdx.x"))
    s[B].vectorize(tx)
    return s, [A, B]

def compute_backward_exp2(dtype, ndim):
    A = tvm.placeholder([tvm.var() for _ in range(ndim)], name='A', dtype=dtype)
    B = tvm.placeholder([tvm.var() for _ in range(ndim)], name='B', dtype=dtype)
    C = tvm.compute([tvm.var() for _ in range(ndim)],
                    lambda *index:  A[index] * B[index] *tvm.const(0.6931471805599453, dtype), name='D')
    s = tvm.create_schedule(C.op)
    return s, A, B, C

@defop(name="backward_exp2", target="cpu", auto_broadcast=True,
       dtype=["float32", "float64"], ndim=list(range(0, 6)))
def backward_exp2(dtype, ndim):
    s, A, B, C = compute_backward_exp2(dtype, ndim)
    # axes = [axis for axis in C.op.axis]
    # fused = s[D].fuse(*axes)
    # s[D].parallel(fused)

    return s, [A, B, C]

@defop(name="cuda_backward_exp2", target="cuda", auto_broadcast=True,
       dtype=["float32", "float64"], ndim=list(range(0, 6)))
def backward_exp2_gpu(dtype, ndim):
    s, A, B, C= compute_backward_exp2(dtype, ndim)
    s = tvm.create_schedule(C.op)
    axes = [axis for axis in C.op.axis]
    fused = s[C].fuse(*axes)
    bx, tx = s[C].split(fused, factor=64)
    s[C].bind(bx, tvm.thread_axis("blockIdx.x"))
    s[C].bind(tx, tvm.thread_axis("threadIdx.x"))
    return s, [A, B, C]


@defop(name="relu", target="cpu", auto_broadcast=True,
       dtype=["float32"], ndim=list(range(0, 6)))
def relu(dtype, ndim):
    A = tvm.placeholder([tvm.var() for _ in range(ndim)], name='A', dtype=dtype)
    B = tvm.compute([tvm.var() for _ in range(ndim)],
                    lambda *index: tvm.if_then_else(A[index] > 0.0, A[index], 0.0), name='B')
    s = tvm.create_schedule(B.op)
    # axes = [axis for axis in B.op.axis]
    # fused = s[B].fuse(*axes)
    # s[B].parallel(fused)
    return s, [A, B]